src/dataflow-types/src/errors.rs

```diff
@@ -40,8 +41,9 @@ impl Codec for DecodeError {
         buf.extend(encoded.iter());
     }
 
-    fn decode<'a>(buf: &'a [u8]) -> Result<Self, String> {
-        let decoded = serde_json::from_slice(buf).map_err(|e| format!("Decoding error: {}", e))?;
+    fn decode<'a>(buf: &'a [u8]) -> Result<Self, CodecError> {
+        let decoded = serde_json::from_slice(buf)
+            .map_err(|e| CodecError::from(format!("Decoding error: {}", e)))?;
```

Comment:
If you provide an `impl From<serde_json::Error> for CodecError` implementation you'll be able to just do `serde_json::from_slice(buf)?;` here. This is because the `?` operator automatically runs `err.into()` on the returned errors so that you can easily use structured errors like you're doing here.

Reply:
I think this pattern is nice and we use it at times in persist, but it shouldn't be applied blindly. In this case, making the change added a dep on serde_json to the persist_types crate, which is bad dep hygene (the impl just converts it into a string; it'd be a different story if CodecError kept the structured serde_json::Error and the dep was avoidable).

The other instances of this review comment are less objectionable since the errors in question in those are stdlib ones.